p2p/trust/trustmetric.go

```diff
@@ -0,0 +1,574 @@
+type TrustMetricConfig struct {
+	// Determines the percentage given to current behavior
+	ProportionalWeight float64
+
+	// Determines the percentage given to prior behavior
+	IntegralWeight float64
+
+	// The window of time that the trust metric will track events across.
+	// This can be set to cover many days without issue
+	TrackingWindow time.Duration
+
+	// Each interval should be short for adapability.
+	// Less than 30 seconds is too sensitive,
+	// and greater than 5 minutes will make the metric numb
+	IntervalLength time.Duration
+}
+
+// DefaultConfig returns a config with values that have been tested and produce desirable results
+func DefaultConfig() TrustMetricConfig {
+	return TrustMetricConfig{
+		ProportionalWeight: 0.4,
+		IntegralWeight:     0.6,
+		TrackingWindow:     (time.Minute * 60 * 24) * 14, // 14 days.
+		IntervalLength:     1 * time.Minute,
+	}
+}
+
+// NewMetric returns a trust metric with the default configuration
+func NewMetric() *TrustMetric {
+	return NewMetricWithConfig(DefaultConfig())
+}
+
+// NewMetricWithConfig returns a trust metric with a custom configuration
+func NewMetricWithConfig(tmc TrustMetricConfig) *TrustMetric {
+	tm := new(TrustMetric)
+	config := customConfig(tmc)
+
+	// Setup using the configuration values
+	tm.proportionalWeight = config.ProportionalWeight
+	tm.integralWeight = config.IntegralWeight
+	tm.intervalLen = config.IntervalLength
+	// The maximum number of time intervals is the tracking window / interval length
+	tm.maxIntervals = int(config.TrackingWindow / tm.intervalLen)
+	// The history size will be determined by the maximum number of time intervals
+	tm.historyMaxSize = intervalToHistoryOffset(tm.maxIntervals) + 1
+	// This metric has a perfect history so far
+	tm.historyValue = 1.0
+	// Setup the channels
+	tm.update = make(chan *updateBadGood, defaultUpdateChanCapacity)
+	tm.trustValue = make(chan *reqTrustValue, defaultRequestChanCapacity)
+	tm.stop = make(chan bool, 2)
+
+	go tm.processRequests()
+	return tm
+}
+
+/* Private methods */
+
+// Ensures that all configuration elements have valid values
+func customConfig(tmc TrustMetricConfig) TrustMetricConfig {
+	config := DefaultConfig()
+
+	// Check the config for set values, and setup appropriately
+	if tmc.ProportionalWeight > 0 {
+		config.ProportionalWeight = tmc.ProportionalWeight
+	}
+
+	if tmc.IntegralWeight > 0 {
+		config.IntegralWeight = tmc.IntegralWeight
+	}
+
+	if tmc.IntervalLength > time.Duration(0) {
+		config.IntervalLength = tmc.IntervalLength
+	}
+
+	if tmc.TrackingWindow > time.Duration(0) &&
+		tmc.TrackingWindow >= config.IntervalLength {
+		config.TrackingWindow = tmc.TrackingWindow
+	}
+
+	return config
+}
+
+// Calculates the derivative component
+func (tm *TrustMetric) derivativeValue() float64 {
+	return tm.proportionalValue() - tm.historyValue
+}
+
+// Strengthens the derivative component when the change is negative
+func (tm *TrustMetric) weightedDerivative() float64 {
```

Comment:
So here `gamma1 = 0` and `gamma2 = 1`, correct? Is that really what we want? And can we make sure it's documented in the proposal.

Reply:
You are correct. These weights have worked well. The result is that no positive change happens when current behavior is >= previous behavior, yet the penalty is at maximum when the current behavior falls below previous behavior. I will include the information in the proposal, and of course we can play with the numbers in the future if necessary.